crates/bevy_ecs/src/term_query/builder.rs

```diff
@@ -0,0 +1,84 @@
+use std::marker::PhantomData;
+
+use crate::{
+    component::ComponentId,
+    prelude::{Component, With, Without, World},
+};
+
+use super::{ComponentTerm, QueryTermGroup, Term, TermQueryState, TermVec};
+
+pub struct QueryBuilder<'w, Q: QueryTermGroup = ()> {
+    terms: TermVec<Term>,
+    current_term: usize,
+    world: &'w mut World,
+    _marker: PhantomData<Q>,
+}
+
+impl<'w, Q: QueryTermGroup> QueryBuilder<'w, Q> {
```

Comment:
This is still missing `Option`, `Has`, `AnyOf`, `Or`, `Added` and `Changed` right?

Reply:
The builder doesn't have helper methods for them but if you can use them in the type signature or add them with `.term::<T>` where T is any `QueryTerm`.

Reply:
Hmm right. Is it useful to have the other helper methods for non-dynamic terms at all? Removing those would simplify things substantially and also help with maintenance.

However, don't we need some way to be able to add them dynamically still?

Reply:
Yeah you're right, they at the very least should have dynamic methods.

There should probably also be some way for you to push a Term you construct to the array to give fine grained access.

The main reason I added helper methods even for the static use case is just because it's a very annoying sequence of keys to type out `term::<With<T>>()` rather than just `with::<T>()`.

Plus I personally find it easier to read.

Reply:
I added dynamic methods for everything except Or and AnyOf as they are composed of other terms and Option as it is a modifier to other terms. I could add booleans everywhere to mark things as optional or have some flag on the builder but neither of those strike me as particularly ergonomic, will think more on it.